[SLP] Initial vectorization of non-power-of-2 ops.

fhahn opened the pull request:
This patch enables vectorization for non-power-of-2 VFs. Initially only
VFs where adding 1 makes the VF a power-of-2, i.e. we can still make
relatively effective use of the vectors.

It relies on the existing target cost-models to return accurate costs for
non-power-of-2 vectors. I checked mostly AArch64 and X86 and
there the costs seem reasonable for the costs I checked, although
I expect there will be a need to refine both the cost-models and lowering
to make most effective use of non-power-of-2 SLP vectorization.

Note that re-ordering and shuffling is not implemented for nodes
requiring padding yet to keep the initial implementation simpler.

The feature is guarded by a new flag, off by defaul for now.

fhahn committed:
[SLP] Initial vectorization of non-power-of-2 ops.

This patch enables vectorization for non-power-of-2 VFs. Initially only
VFs where adding 1 makes the VF a power-of-of-2, i.e. we can still make
relatively effective use of the vectors.

It relies on the existing target cost-models to return accurate costs for
non-power-of-2 vectors. I checked mostly AArch64 and X86 and
there the costs seem reasonable for the costs I checked, although
I expect there will be a need to refine both the cost-models and lowering
to make most effective use of non-power-of-2 SLP vectorization.

Note that re-ordering and shuffling is not implemented for nodes
requiring padding yet to keep the initial implementation simpler.

The feature is guarded by a new flag, off by defaul for now.

alexey-bataev commented:
Thanks for the patch, but this is too early to land. I'm working on non-power-of-2 vectorization (it is WIP and still requires couple more patches to go). Being implemented without some extra patches it leads to perf regression in some cases. Need to handle all this stuff correctly.
Would be good if you could land new tests separately, may help with the perf gains/regressions later.

fhahn commented:
> Thanks for the patch, but this is too early to land. I'm working on non-power-of-2 vectorization (it is WIP and still requires couple more patches to go). Being implemented without some extra patches it leads to perf regression in some cases. Need to handle all this stuff correctly. 

Do you have any more info about where those perf regressions are showing up (like architecture, benchmark, code pattern)? Curious if they are mostly related to reordering/shuffling/improved gathering? This patch does only support full gather, which should hopefully reflect the cost quite accurately (modulo cost-model gaps)

> Would be good if you could land new tests separately, may help with the perf gains/regressions later.

Done! Some of the tests are for profitability, but many of them are also to cover crashes.

alexey-bataev commented:
> > Thanks for the patch, but this is too early to land. I'm working on non-power-of-2 vectorization (it is WIP and still requires couple more patches to go). Being implemented without some extra patches it leads to perf regression in some cases. Need to handle all this stuff correctly.
> 
> Do you have any more info about where those perf regressions are showing up (like architecture, benchmark, code pattern)? Curious if they are mostly related to reordering/shuffling/improved gathering? This patch does only support full gather, which should hopefully reflect the cost quite accurately (modulo cost-model gaps)

I investigated Spec + some other benchmarks, mostly the regressions are because of the too early SLP vectorization with LTO scenarios. To fix this, need to implement couple more patches (vectorization of gathered loads + reordering) + need to add some limitations.
The reordering is also a problem, but not the biggest one.
This new node will be a burden to support, later the genric Vectorization will be extended to support non-power-of-2 nodes directly.

> 
> > Would be good if you could land new tests separately, may help with the perf gains/regressions later.
> 
> Done! Some of the tests are for profitability, but many of them are also to cover crashes.

fhahn commented:
> This new node will be a burden to support, later the genric Vectorization will be extended to support non-power-of-2 nodes directly.

Agreed, it turns out the new node is not really needed and non-power-of-2 vector ops are handled quite well directly by backends (checked AArch64). Adjusted the patch to drop the special TreeEntry, tracking of NumPadding and generate non-power-of-2 vector ops directly for all nodes in the tree, WDYT as a starting point? https://github.com/llvm/llvm-project/commit/0ddbdd3052e694eedef6bccce3b17f92dff7add7

alexey-bataev commented:
As I said before, landing this alone causes perf regressions. I'm working on this, need to land couple patches to avoid this. I'm working on this feature for 3 years already (or more), most part of my previous patches are related to non-power-of-2 support.

fhahn commented:
> As I said before, landing this alone causes perf regressions. I'm working on this, need to land couple patches to avoid this. I'm working on this feature for 3 years already (or more), most part of my previous patches are related to non-power-of-2 support.

I did perf evaluation for this change on SPEC2017 and some large internal benchmarks on AArch64, and the only regressions I was able to spot were due to AArch64 cost-model issues, where the cost of non-power-of-2 vector ops wasn't estimated accurately (one instance fixed by #78181). 

With the initial level of support, do you think there are fundamental issues in SLPVectorizer that aren't related to target specific cost models not providing accurate costs? To iterate on the latter, having limited support early would be helpful as people can fix their cost models early and also iteratively as more features enabled for non-power-of-2 vec.

alexey-bataev commented:
> > As I said before, landing this alone causes perf regressions. I'm working on this, need to land couple patches to avoid this. I'm working on this feature for 3 years already (or more), most part of my previous patches are related to non-power-of-2 support.
> 
> I did perf evaluation for this change on SPEC2017 and some large internal benchmarks on AArch64, and the only regressions I was able to spot were due to AArch64 cost-model issues, where the cost of non-power-of-2 vector ops wasn't estimated accurately (one instance fixed by #78181).
> 
> With the initial level of support, do you think there are fundamental issues in SLPVectorizer that aren't related to target specific cost models not providing accurate costs? To iterate on the latter, having limited support early would be helpful as people can fix their cost models early and also iteratively as more features enabled for non-power-of-2 vec.

It contradicts with my work, I did for several years. I have a plan of implementing this feature and moving forward step by step. Most of this newly added stuff, like padding etc., is not needed, if everything is implemented correctly. This is not a complete solution, but just a small hack to enable the feature, which may cause issues in future and adds extra burden to maintaining and removing it later.

fhahn commented:
> Most of this newly added stuff, like padding etc., is not needed, if everything is implemented correctly. This is not a complete solution, but just a small hack to enable the feature, which may cause issues in future and adds extra burden to maintaining and removing it later.

Agreed that the padding isn't needed, did you see 0ddbdd3052e694eedef6bccce3b17f92dff7add7 mentioned in one of  my previous comment that removes all the padding stuff? 

There are still a few things that can be cleaned up there, but with that removed, the changes mostly boil down to disabling shuffling/reordering, as most of that code isn't able to handle non-power-of-2 vectors yet. As support is added for those, the restrictions could be loosened gradually.

fhahn commented on llvm/test/Transforms/SLPVectorizer/X86/vect_copyable_in_binops.ll:
```diff
@@ -668,21 +702,32 @@ entry:
+; NON-POW2-NEXT:    store float [[TMP0]], ptr [[DST]], align 4
+; NON-POW2-NEXT:    [[TMP1:%.*]] = load <3 x float>, ptr [[INCDEC_PTR]], align 4
+; NON-POW2-NEXT:    [[TMP2:%.*]] = fadd <3 x float> [[TMP1]], <float 1.000000e+00, float 2.000000e+00, float 3.000000e+00>
+; NON-POW2-NEXT:    store <3 x float> [[TMP2]], ptr [[INCDEC_PTR1]], align 4
```
(Moved comment  here https://github.com/llvm/llvm-project/commit/0ddbdd3052e694eedef6bccce3b17f92dff7add7#r137302312)
> Is this supported on all platforms? Or you rely on the check that this will end up with the scalarized version in the end?
 
I checked on AArch64 and X86 and both support OK lowering of such operations, although in a number of cases improvements can be made, e.g. #78632 and #78637 for loads and stores of <3 x i8> on AArch64.

This effectively relies on the cost-models to return accurate costs for non-power-of-2 vector operations. If a vector op of a non-power-of-2 type will get scalarizied in the backend, the cost model should return a high cost.

fhahn reviewed (commented):
Updated the PR with the version without padding, including addressing @alexey-bataev 's comments at 0ddbdd3052e694eedef6bccce3b17f92dff7add7

fhahn commented:
Removed the remaining references to padding, further simplified changes and removed additional unneeded changes.

Ping :)

fhahn committed:
Remove stale PADDING check lines, fix POW2/NON-POW2 prefixes in test.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -2733,6 +2737,9 @@ class BoUpSLP {
                           SmallVectorImpl<Value *> *OpScalars = nullptr,
                           SmallVectorImpl<Value *> *AltScalars = nullptr) const;
 
+    /// Return the number of padding lanes (containg poison) for this node.
```
```suggestion
    /// Return the number of padding lanes (containing poison) for this node.
```

RKSimon commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -2891,9 +2898,13 @@ class BoUpSLP {
-
+      if (!isPowerOf2_32(Last->Scalars.size())) {
+        assert((Last->ReorderIndices.empty()) &&
+               "Reodering isn't implemented for nodes with padding yet");
```
Reodering ?

fhahn commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -13555,9 +13619,39 @@ bool SLPVectorizerPass::vectorizeStores(ArrayRef<StoreInst *> Stores,
+          }
+        }
+      }
+
```
One alternative to make it part of the loop would be to collect VFs up front in a vector, which I did in the latest version, WDYT?. Could also extract to a lambda and keep the NonPowerOf2 checks as in the original patch.

fhahn commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -10648,7 +10699,16 @@ Value *BoUpSLP::vectorizeOperand(TreeEntry *E, unsigned NodeIdx,
 template <typename BVTy, typename ResTy, typename... Args>
 ResTy BoUpSLP::processBuildVector(const TreeEntry *E, Args &...Params) {
   assert(E->State == TreeEntry::NeedToGather && "Expected gather node.");
+
```
Done, thanks!

fhahn committed:
!fixup Address latest comments, thanks!

fhahn reviewed (commented):
Addressed comments, thanks!

fhahn committed:
!fixup  Add fixme to processBuildVector

also use {} instead of (empty) ResueShuffleIndices.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -6987,6 +7024,17 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
     auto *VecTy = FixedVectorType::get(VL.front()->getType(), VL.size());
     InstructionCost GatherCost = 0;
     SmallVector<Value *> Gathers(VL.begin(), VL.end());
+    auto ComputeGatherCost = [&]() {
+      return all_of(Gathers, UndefValue::classof)
+                 ? TTI::TCC_Free
+                 : R.getGatherCost(Gathers, !Root && VL.equals(Gathers));
+    };
```
Why the default code path does not work here?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -10654,7 +10703,16 @@ template <typename BVTy, typename ResTy, typename... Args>
+    SmallVector<int> Mask(VF, PoisonMaskElem);
+    std::iota(Mask.begin(), Mask.end(), 0);
+    ShuffleBuilder.add(BV, Mask);
+    return ShuffleBuilder.finalize({});
```
```suggestion
    return ShuffleBuilder.finalize(std::nullopt);
```

fhahn replied:
Adusted, thanks

fhahn committed:
Merge branch 'main' into slp-vec3

fhahn committed:
!fixup undo gather cos changes.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -179,6 +179,10 @@ static cl::opt<bool>
 
+static cl::opt<bool> VectorizeNonPowerOf2(
+    "slp-vectorize-non-power-of-2", cl::init(false), cl::Hidden,
+    cl::desc("Try to vectorize with non-power-of-2 with number of elements."));
```
```suggestion
    cl::desc("Try to vectorize with non-power-of-2 number of elements."));
```

fhahn replied:
Dropped, thanks!

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -10661,7 +10704,16 @@ template <typename BVTy, typename ResTy, typename... Args>
   assert(E->State == TreeEntry::NeedToGather && "Expected gather node.");
   unsigned VF = E->getVectorFactor();
-
+  BVTy ShuffleBuilder(Params...);
+  // FIXME: Only full gathering is implemented for non-power-of-2 nodes at the
+  // moment.
+  if (E->isNonPowOf2Vec()) {
+    Value *BV = ShuffleBuilder.gather(E->Scalars);
+    SmallVector<int> Mask(VF, PoisonMaskElem);
+    std::iota(Mask.begin(), Mask.end(), 0);
+    ShuffleBuilder.add(BV, Mask);
+    return ShuffleBuilder.finalize(std::nullopt);
+  }
```
Again, why do you need special processing for non-power-of-2? It would be good, if you could try to adjust this whole function to support it. May help in future for the major non-power-of-2 functionality.

fhahn replied:
It's not needed in the latest version, removed, thanks!

alexey-bataev reviewed (commented):
Would be good to see more tests for other targets. May help to check that TTI does not crash and vectorizatyion works for these targets too.

fhahn committed:
Merge remote-tracking branch 'origin/main' into slp-vec3

fhahn committed:
!fixup remove escape hatch for non-power-of-2 vectors from processBV.

fhahn committed:
!fixup removed with in wrong place

fhahn reviewed (commented):
> Would be good to see more tests for other targets. May help to check that TTI does not crash and vectorizatyion works for these targets too.

Excellent point, I added a number of X86 tests (check lines updated in the latest version of the PR) and also added extra X86 for issues I found while testing on X86. AArch64 and X86 are the architectures I can test locally, are there any particular other targets you think would be good to have tests for ?

fhahn committed:
!fixup also update odd_store.ll

fhahn committed:
Merge branch 'main' into slp-vec3

fhahn committed:
Merge remote-tracking branch 'origin/main' into slp-vec3

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
[SLP] Collect candidate VFs in vector in vectorizeStores (NFC).

This is in preparation for
https://github.com/llvm/llvm-project/pull/77790 and makes it easy to add
other, non-power-of-2 VFs for processing.

fhahn committed:
Merge branch 'users/fhahn/slp-store-vfs-in-vector' into slp-vec3

fhahn committed:
!fixup update tests after merge.

fhahn committed:
Merge remote-tracking branch 'origin/main' into users/fhahn/slp-store-vfs-in-vector

fhahn committed:
[SLP] Exit early if MaxVF < MinVF (NFCI).

Exit early if MaxVF < MinVF. In that case, the loop body below will
never get entered. Note that this adjusts the condition from MaxVF <=
MinVF. If MaxVF == MinVF, vectorization may still be feasible (and the
loop below gets entered).

fhahn committed:
[SLP] Exit early .

fhahn committed:
Merge remote-tracking branch 'origin/users/fhahn/slp-early-exit' into users/fhahn/slp-store-vfs-in-vector

fhahn committed:
!fixup use for_each.

fhahn committed:
Merge remote-tracking branch 'origin/users/fhahn/slp-store-vfs-in-vector' into slp-vec3

	llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

fhahn committed:
Merge remote-tracking branch 'origin/main' into users/fhahn/slp-store-vfs-in-vector

fhahn committed:
Merge remote-tracking branch 'origin/users/fhahn/slp-store-vfs-in-vector' into slp-vec3

fhahn committed:
!fixup add non-power-of-2 VF correctly.

fhahn reviewed (commented):
Updated to add non-power-of-2 VF based on https://github.com/llvm/llvm-project/pull/82793

fhahn committed:
Merge remote-tracking branch 'origin/main' into slp-vec3

fhahn commented:
ping :)

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -7468,7 +7508,8 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
             continue;
-          GatherCost += getBuildVectorCost(VL.slice(I, VF), Root);
+          GatherCost += getBuildVectorCost(
+              VL.slice(I, std::min<unsigned>(VL.size() - I, VF)), Root);
```
```suggestion
              VL.slice(I, std::min(End - I, VF)), Root);
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -11088,7 +11132,6 @@ template <typename BVTy, typename ResTy, typename... Args>
 ResTy BoUpSLP::processBuildVector(const TreeEntry *E, Args &...Params) {
   assert(E->State == TreeEntry::NeedToGather && "Expected gather node.");
   unsigned VF = E->getVectorFactor();
-
```
Restore this line

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -14036,10 +14084,22 @@ bool SLPVectorizerPass::vectorizeStores(ArrayRef<StoreInst *> Stores,
 
       unsigned Sz = 1 + Log2_32(MaxVF) - Log2_32(MinVF);
       SmallVector<unsigned> CandidateVFs(Sz);
+      auto VFsToFill = make_range(CandidateVFs.begin(), CandidateVFs.end());
```
Why need to use VFsToFill, why CandidateVFs cannot be used directly?

alexey-bataev replied:
Why need to try it first?

fhahn replied:
If we start with the regular power-of-2 VFs first, then in many cases we vectorize with a power-of-2 VF smaller than the non-power-of-2 one eagerly.

alexey-bataev replied:
Something like 
```
SmallVector<unsigned> CandidateVFs(Sz+(non-power-of-2-vf ? 1 : 0));
unsigned Size = MinVF;
for_each(reverse(CandidateVFs), [&](unsigned &VF) {
        VF = Size > MaxVF ? non-power-of-2-vf : Size;
        Size *= 2;
      });
```
?

fhahn committed:
Merge remote-tracking branch 'origin/main' into slp-vec3

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
!fixup fix formatting

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -5748,7 +5775,8 @@ BoUpSLP::TreeEntry::EntryState BoUpSLP::getScalarsVectorizationState(
   case Instruction::ExtractElement: {
     bool Reuse = canReuseExtract(VL, VL0, CurrentOrder);
-    if (Reuse || !CurrentOrder.empty())
+    // FIXME: Vectorizing is not supported yet for non-power-of-2 ops.
+    if (isPowerOf2_32(VL.size()) && (Reuse || !CurrentOrder.empty()))
```
```suggestion
    // FIXME: Vectorizing is not supported yet for non-power-of-2 ops.
    if (!isPowerOf2_32(VL.size())
      return TreeEntry::NeedToGather;
    bool Reuse = canReuseExtract(VL, VL0, CurrentOrder);
    if (Reuse || !CurrentOrder.empty())
```

fhahn replied:
Adjusted, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into slp-vec3

fhahn committed:
!fixup add separate early exit for Non-power-of-2 VFs.

fhahn committed:
Merge remote-tracking branch 'origin/main' into slp-vec3

fhahn committed:
!fixup adjust VF computation as suggested

fhahn committed:
Merge branch 'main' into slp-vec3

fhahn commented:
Ping :)

fhahn commented:
ping

fhahn committed:
Merge branch 'main' into slp-vec3

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -6105,6 +6135,12 @@ void BoUpSLP::buildTree_rec(ArrayRef<Value *> VL, unsigned Depth,
           if (PWSz == VL.size()) {
             ReuseShuffleIndicies.clear();
           } else {
+            if (UserTreeIdx.UserTE && UserTreeIdx.UserTE->isNonPowOf2Vec()) {
+              LLVM_DEBUG(dbgs() << "SLP: Reshuffling scalars not yet supported "
+                                   "for nodes with padding.\n");
+              newTreeEntry(VL, std::nullopt /*not vectorized*/, S, UserTreeIdx);
+              return false;
+            }
```
Better to move this check and one below to line (or after) 6399 in the original source code to avoid duplication.

fhahn replied:
I moved the check up the beginning of the `else { ` in the lambada to avoid the duplication. I think it cannot be moved out of `TryToFindDuplicates`, as we only need to have the bail out when reshuffling would be needed.

fhahn committed:
Merge remote-tracking branch 'origin/main' into slp-vec3

fhahn committed:
!fixup address comments, update after upstream changes.

fhahn committed:
!fixup remove newline

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -5805,7 +5832,10 @@ BoUpSLP::TreeEntry::EntryState BoUpSLP::getScalarsVectorizationState(
+    // FIXME: Vectorizing is not supported yet for non-power-of-2 ops.
+    if (!isPowerOf2_32(VL.size()))
+      return TreeEntry::NeedToGather;
+    if ((Reuse || !CurrentOrder.empty()))
```
Remove extra parens, restore original check here

fhahn replied:
Done, thanks!

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -6111,6 +6141,12 @@ void BoUpSLP::buildTree_rec(ArrayRef<Value *> VL, unsigned Depth,
     if (NumUniqueScalarValues == VL.size()) {
       ReuseShuffleIndicies.clear();
     } else {
+      if (UserTreeIdx.UserTE && UserTreeIdx.UserTE->isNonPowOf2Vec()) {
```
Add FIXME here for non-power-of-2 support

fhahn replied:
Added, thanks!

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -14798,14 +14843,23 @@ bool SLPVectorizerPass::vectorizeStores(ArrayRef<StoreInst *> Stores,
         continue;
       }
 
+      std::optional<unsigned> NonPowerOf2VF;
```
```suggestion
      unsigned NonPowerOf2VF = 0;
```

fhahn replied:
Done, thanks!

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -14798,14 +14843,23 @@ bool SLPVectorizerPass::vectorizeStores(ArrayRef<StoreInst *> Stores,
-      for_each(CandidateVFs, [&](unsigned &VF) {
-        VF = Size;
-        Size /= 2;
+      SmallVector<unsigned> CandidateVFs(Sz + bool(NonPowerOf2VF));
```
```suggestion
      SmallVector<unsigned> CandidateVFs(Sz + (NonPowerOf2VF > 0 ? 1 : 0));
```
Better to avoid adding `bool`

fhahn replied:
Done, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into slp-vec3

fhahn commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -14798,14 +14843,23 @@ bool SLPVectorizerPass::vectorizeStores(ArrayRef<StoreInst *> Stores,
+        unsigned CandVF = Operands.size();
+        if (isPowerOf2_32(CandVF + 1) && CandVF <= MaxVF) {
+          NonPowerOf2VF = CandVF;
+        }
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -14798,14 +14843,23 @@ bool SLPVectorizerPass::vectorizeStores(ArrayRef<StoreInst *> Stores,
+      SmallVector<unsigned> CandidateVFs(Sz + bool(NonPowerOf2VF));
+      unsigned Size = MinVF;
+      for_each(reverse(CandidateVFs), [&](unsigned &VF) {
+        VF = Size > MaxVF ? *NonPowerOf2VF : Size;
```
Done, thanks!

fhahn committed:
!fixup address latest comments, thanks!

fhahn reviewed (commented):
Latest comments should be addressed, thanks!

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -2806,6 +2810,9 @@ class BoUpSLP {
                           SmallVectorImpl<Value *> *AltScalars = nullptr) const;
 
+    /// Return true if this is a non-power-of-2 node.
+    bool isNonPowOf2Vec() const { return !isPowerOf2_32(Scalars.size()); }
```
What if ReuseShuffleIndices is not empty? Will it work?Can you add a test?

fhahn replied:
All code paths should guard against that AFAICT. I added an assertion to make sure. Couldn't find any test case that triggers this  across large code bases (SPEC2006, SPEC2017, llvm-test-suite, clang bootstrap and large internal benchmarks)

fhahn committed:
Merge remote-tracking branch 'origin/main' into slp-vec3

fhahn committed:
!fixup add assert

fhahn commented:
ping

alexey-bataev reviewed (approved):
LG

fhahn committed:
Merge branch 'main' into slp-vec3

dtcxzyw commented:
@fhahn This PR breaks our CI: https://github.com/dtcxzyw/llvm-opt-benchmark/issues/514

```
opt: /home/dtcxzyw/WorkSpace/Projects/compilers/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:169: const T& llvm::ArrayRef<T>::front() const [with T = llvm::Value*]: Assertion `!empty()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: bin/opt -passes=slp-vectorizer reduced.ll
 #0 0x0000721336bed7b0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.19.0git+0x1ed7b0)
 #1 0x0000721336beabbf llvm::sys::RunSignalHandlers() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.19.0git+0x1eabbf)
 #2 0x0000721336bead15 SignalHandler(int) Signals.cpp:0:0
 #3 0x0000721336642520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007213366969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007213366969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007213366969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x0000721336642476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007213366287f3 abort ./stdlib/abort.c:81:7
 #9 0x000072133662871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x0000721336639e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x000072133178328c llvm::SLPVectorizerPass::vectorizeStores(llvm::ArrayRef<llvm::StoreInst*>, llvm::slpvectorizer::BoUpSLP&)::'lambda'(std::set<std::pair<unsigned int, int>, llvm::SLPVectorizerPass::vectorizeStores(llvm::ArrayRef<llvm::StoreInst*>, llvm::slpvectorizer::BoUpSLP&)::StoreDistCompare, std::allocator<std::pair<unsigned int, int>>> const&)::operator()(std::set<std::pair<unsigned int, int>, llvm::SLPVectorizerPass::vectorizeStores(llvm::ArrayRef<llvm::StoreInst*>, llvm::slpvectorizer::BoUpSLP&)::StoreDistCompare, std::allocator<std::pair<unsigned int, int>>> const&) const SLPVectorizer.cpp:0:0
#12 0x0000721331783b90 llvm::SLPVectorizerPass::vectorizeStores(llvm::ArrayRef<llvm::StoreInst*>, llvm::slpvectorizer::BoUpSLP&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMVectorize.so.19.0git+0x183b90)
#13 0x0000721331784286 llvm::SLPVectorizerPass::vectorizeStoreChains(llvm::slpvectorizer::BoUpSLP&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMVectorize.so.19.0git+0x184286)
#14 0x000072133178540f llvm::SLPVectorizerPass::runImpl(llvm::Function&, llvm::ScalarEvolution*, llvm::TargetTransformInfo*, llvm::TargetLibraryInfo*, llvm::AAResults*, llvm::LoopInfo*, llvm::DominatorTree*, llvm::AssumptionCache*, llvm::DemandedBits*, llvm::OptimizationRemarkEmitter*) (.part.0) SLPVectorizer.cpp:0:0
#15 0x0000721331785e1b llvm::SLPVectorizerPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMVectorize.so.19.0git+0x185e1b)
#16 0x000072133206f636 llvm::detail::PassModel<llvm::Function, llvm::SLPVectorizerPass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMPasses.so.19.0git+0x6f636)
#17 0x000072132f51eec1 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x31eec1)
#18 0x00007213358a45d6 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.19.0git+0xa45d6)
#19 0x000072132f51db9b llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x31db9b)
#20 0x00007213358a59c6 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.19.0git+0xa59c6)
#21 0x000072132f51b9d1 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x31b9d1)
#22 0x000072133702e5c5 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.19.0git+0x245c5)
#23 0x0000721337039216 optMain (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.19.0git+0x2f216)
#24 0x0000721336629d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#25 0x0000721336629e40 call_init ./csu/../csu/libc-start.c:128:20
#26 0x0000721336629e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#27 0x0000566543d65095 _start (bin/opt+0x1095)
Aborted (core dumped)
```

Reduced test case:
```
; opt -passes=slp-vectorizer reduced.ll -S

target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define void @test(i24 %0, ptr %.pre) {
  br label %2

2:                                                ; preds = %2, %1
  %.sroa.28.0.extract.trunc = trunc i24 %0 to i8
  store i8 %.sroa.28.0.extract.trunc, ptr %.pre, align 1
  %3 = getelementptr i8, ptr %.pre, i64 1
  store i8 0, ptr %3, align 1
  br label %2
}
```

fhahn commented:
@dtcxzyw thanks for the reproducer, the issue was that in some cases on X86, MinVF wasn't a power-of-2. Fixed by using the stored type to compute the VF passed to `getStoreMinimumVF`
